### PR TITLE
Update dependencies to kubernetes-1.9.1

### LINKED ISSUE #1610: Rebase on k8s 1.9.1 dependencies
One major motivator for this is that we will be able to use explicit versions of these dependenceies that are human-readable and consistent (`kubernetes-1.8`) instead of a random collection of shas.

### PR DESCRIPTION
**Pull request by nilebox**
Resolves #1610
As usual, review individual commits and skip the commit with vendoring deps.
Manual changes:
- make conflicting dependencies in-sync with k8s.io/apiserver and k8s.io/kubernetes
- switch to runtime.Object.DeepCopy()
- follow-up on admission plugins refactoring
- etcd/apiserver config changes

### CODE REVIEW

**Comment by nilebox** [comment]
//TODO: we have copied some k8s.io/kubernetes code to our codebase (to avoid depending on entire k8s.io/kubernetes repo), in a follow-up PR it would be nice to check if there were any changes upstream.

**Comment by nilebox** [comment]
@arschles the CI log is full of messages like
```
E1219 22:27:11.688959      68 controller_instance.go:1611] ServiceInstance "test-namespace/test-instance": Failed to update status: Operation cannot be fulfilled on serviceinstances.servicecatalog.k8s.io "test-instance": the object has been modified; please apply your changes to the latest version and try again
```
Not entirely clear which change caused this, but I suspect there's something wrong with `resourceVersion`/`generation` properly set after `DeepCopy()`.

**Comment by nilebox** [comment]
The final error
```
==4456==ERROR: ThreadSanitizer failed to allocate 0x80000 (524288) bytes of DenseSlabAllocator (error code: 12)

FATAL: ThreadSanitizer CHECK failed: ./gotsan.cc:6532 "((0 && "unable to mmap")) != (0)" (0x0, 0x0)
```
is "interesting" tho :)

File: glide.yaml [hunk 1]
```diff
@@ -95,4 +91,4 @@ import:
 - package: github.com/pmorie/go-open-service-broker-client
   version: 31d8027f493f8f23f850415d171c7c52a972a6f2
 - package: github.com/coreos/etcd
-  version: ^3.2.9
```
Comment (MHBauer) [inline glide.yaml]: Not sure this version of etcd will let us do the embedded server trick.

File: glide.yaml [hunk 2]
```diff
@@ -66,27 +66,23 @@ import:
...
+  version: kubernetes-1.9.0 # tag/kubernetes-1.9.0
 - package: k8s.io/kube-openapi
-  version: 868f2f29720b192240e18284659231b440f9cda5 # k8s.io 1.8 (Oct 4, 2017)
+  version: 39a7bf85c140f972372c2a0d1ee40adbf0c8bfe1 # in sync with https://github.com/kubernetes/kubernetes/blob/v1.9.0/Godeps/Godeps.json
 - package: k8s.io/api
-  version: fe29995db37613b9c5b2a647544cf627bfa8d299 # release-1.8, latest (Oct 4, 2017)
+  version: kubernetes-1.9.0 # tag/kubernetes-1.9.0
 - package: k8s.io/apiserver
-  version: f9b476d61dbdf6474e3d42cc5e854d0d8b7f759e # release-1.8, latest (Oct 4, 2017)
+  version: kubernetes-1.9.0 # tag/kubernetes-1.9.0
```
Comment (MHBauer) [inline glide.yaml]: Yay, 1.9 versions!

File: pkg/apis/servicecatalog/serialization_test.go [hunk 3]
```diff
@@ -114,11 +114,7 @@ func roundTrip(t *testing.T, codec runtime.Codec, item runtime.Object) {
 	printer := spew.ConfigState{DisableMethods: true}
 
 	original := item
-	copied, err := api.Scheme.DeepCopy(item)
-	if err != nil {
-		panic(fmt.Sprintf("unable to copy: %v", err))
-	}
-	item = copied.(runtime.Object)
+	item = item.DeepCopyObject()
```
Comment (MHBauer) [inline pkg/apis/servicecatalog/serialization_test.go]: Neat to see this in action.

File: pkg/registry/servicecatalog/instance/storage.go [hunk 4]
```diff
@@ -207,6 +207,6 @@ func (r *ReferenceREST) Get(ctx genericapirequest.Context, name string, options
 
 // Update alters the reference subset of an object and it
 // implements rest.Updater interface
-func (r *ReferenceREST) Update(ctx genericapirequest.Context, name string, objInfo rest.UpdatedObjectInfo) (runtime.Object, bool, error) {
-	return r.store.Update(ctx, name, objInfo)
+func (r *ReferenceREST) Update(ctx genericapirequest.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc) (runtime.Object, bool, error) {
+	return r.store.Update(ctx, name, objInfo, createValidation, updateValidation)
```
Comment (MHBauer) [inline pkg/registry/servicecatalog/instance/storage.go]: wow, it looks like the update interface made a big change.

File: glide.yaml (hunk 2 above)
Comment (nilebox) [reply glide.yaml]: Yep, finally we have tags in non-core repos.

File: glide.yaml (hunk 1 above)
Comment (MHBauer) [reply glide.yaml]: I'm thinking 3.0.10

Comment (nilebox) [reply glide.yaml]: Interesting. I was not aware of the features we rely on in a particular version of etcd.
`3.1.10` is the version defined by `k8s.io/kubernetes`, so I just wanted it to be consistent.
If we do require `3.2.9+`, I can revert this change of course.

**Comment by MHBauer** [comment]
Kicked travis to see if this memory thing is consistent. What could be doing that requires so much memory?

**Comment by MHBauer** [comment]
There's definitely some kind of massive memory leak locally for me. I had 4gig+ of memory used at one point.

**Comment by MHBauer** [comment]
Current guess involves the embedded etcd. The integration test has about a 1000 open connections before it just hangs.

output of `lsof` is filled with
```
integrati 16208 16264             root   12u     IPv4           31705855      0t0        TCP localhost:43420->localhost:2379 (ESTABLISHED)
...
integrati 16208 16264             root  904u     IPv4           31722577      0t0        TCP localhost:46726->localhost:2379 (ESTABLISHED)
```

**Comment by n3wscott** [comment]
etcd should be shut down at the end of each test.

**Comment by MHBauer** [comment]
> etcd should be shut down at the end of each test.
@n3wscott As in it is not already, or as in it is failing to do so?

**Comment by MHBauer** [comment]
That button should really be bright red.

**Comment by nilebox** [comment]
The log messages about failing updates I wrote about in https://github.com/kubernetes-incubator/service-catalog/pull/1633#issuecomment-352906819 seems to be caused https://github.com/kubernetes-incubator/service-catalog/issues/1639 and they shouldn't fail the build.

So the main issue with the Travis build seems to be caused by crashing etcd.

**Comment by nilebox** [comment]
Tried bumping `etcd` and `grpc` versions (to see if it will perform better), but found the same issue with watch as reported in https://github.com/kubernetes/kubernetes/pull/57480#issuecomment-353389046. After this Kubernetes PR is fixed, we can bump our dependencies as well.

**Comment by nilebox** [comment]
@MHBauer Apparently bumping an `etcd` version helped to fix the Travis CI build, but as I mentioned above there is some issue with watch introduced in this version of etcd or grpc, so I would wait for the fix.

Alos, the fact that our CI still passes suggests that we don't properly test watch functionality?

**Comment by MHBauer** [comment]
I'm not sure there's a layer we can know if watch works. Don't the watchers fall back to polling if they aren't available? Is there a way to know?

**Comment by arschles** [comment]
@MHBauer @nilebox what's the state of this PR? any idea what was going on with the OOM-ish failure? have there been enough consecutive CI successes to deem it reliable? other things I'm not considering?

**Comment by nilebox** [comment]
@arschles I believe there are no left issues with this PR. I resolved conflicts and bumped k8s dependencies to `1.9.1` and etcd `3.2.13` (the one that is being used in `k8s/kubernetes` master), and the integration build succeeded in Jenkins. Travis build hasn't been run yet as they are experiencing some issues with their build agents (see https://www.traviscistatus.com/)

@arschles @MHBauer can you LGTM this PR if there are no more comments on it?

**Comment by nilebox** [comment]
Travis build has succeeded as well.

**Comment by pmorie** [comment]
LGTM, thanks @nilebox

**Comment by arschles** [comment]
also @nilebox can you remove the `do-not-merge` label if appropriate?

**Comment by MHBauer** [comment]
I'll run a `make test` on a merged base and look at the result.

**Pull request merged**